### PR TITLE
USR-2: Personal Account

### PR DESCRIPTION
**Pull request by Johnny7c3**

### CODE REVIEW

File: src/main/java/ru/jma/userservice/controller/UserController.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.web.bind.annotation.*;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import ru.jma.userservice.model.User;
+import ru.jma.userservice.service.UserService;
+
+@RestController
+public class UserController {
```
Comment (FanilBiktimirov) [inline src/main/java/ru/jma/userservice/controller/UserController.java:13]: requestMapping = "/users"

Comment (FanilBiktimirov) [inline src/main/java/ru/jma/userservice/controller/UserController.java:13]: Add lombok annotation @RequiredArgsConstructor

File: src/main/java/ru/jma/userservice/controller/UserController.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+    private UserService userService;
```
Comment (ilya2236) [inline src/main/java/ru/jma/userservice/controller/UserController.java]: Make this field final

File: src/main/java/ru/jma/userservice/model/User.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
+package ru.jma.userservice.model;
+
+import lombok.*;
+import org.springframework.data.annotation.Id;
+import org.springframework.data.relational.core.mapping.Table;
+
+@Table
+@Data
+@NoArgsConstructor
```
Comment (FanilBiktimirov) [inline src/main/java/ru/jma/userservice/model/User.java:20]: Remove @AllArgsConstructor, @EqualsAndHashCode

File: src/main/java/ru/jma/userservice/controller/UserController.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+    @Autowired
+    public UserController(UserService userService) {
+        this.userService = userService;
+    }
+
+    @GetMapping("/users")
+    public Flux<User> getAll() {
```
Comment (FanilBiktimirov) [inline src/main/java/ru/jma/userservice/controller/UserController.java]: Add user suffix in all controller methods

File: src/main/java/ru/jma/userservice/repository/UserRepository.java [hunk 5]
```diff
@@ -0,0 +1,6 @@
+package ru.jma.userservice.repository;
+
+import org.springframework.data.repository.reactive.ReactiveCrudRepository;
+import ru.jma.userservice.model.User;
+
+public interface UserRepository extends ReactiveCrudRepository<User,Long> {}
```
Comment (FanilBiktimirov) [inline src/main/java/ru/jma/userservice/repository/UserRepository.java]: add space

Comment (FanilBiktimirov) [reply src/main/java/ru/jma/userservice/repository/UserRepository.java]: ctrl+alt+L

File: src/main/java/ru/jma/userservice/service/UserService.java [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import ru.jma.userservice.model.User;
+import ru.jma.userservice.repository.UserRepository;
+
+@Service
+public class UserService {
```
Comment (FanilBiktimirov) [inline src/main/java/ru/jma/userservice/service/UserService.java:12]: @RequiredArgsConstructor

File: src/main/java/ru/jma/userservice/service/UserService.java [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+    private UserRepository userRepository;
```
Comment (ilya2236) [inline src/main/java/ru/jma/userservice/service/UserService.java]: Make this field final

File: src/main/java/ru/jma/userservice/service/UserService.java [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+    @Autowired
+    public UserService(UserRepository userRepository) {
+        this.userRepository = userRepository;
+    }
+
+    public Flux<User> findAll() {
```
Comment (FanilBiktimirov) [inline src/main/java/ru/jma/userservice/service/UserService.java]: Add user suffix in all services methods and prefix get

File: src/main/java/ru/jma/userservice/service/UserService.java [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+    public Mono<User> findById(Long id) {
+        return userRepository.findById(id);
+    }
+
+    public Mono<User> save(User user) {
+        return userRepository.save(user);
+    }
+
+    public Mono<User> update(Long id, User user) {
+        return findById(id).map(animeToUpdate -> user.withId(animeToUpdate.getId()))
```
Comment (FanilBiktimirov) [inline src/main/java/ru/jma/userservice/service/UserService.java]: naming

File: src/main/java/ru/jma/userservice/service/UserService.java [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+
+    public Mono<User> findById(Long id) {
+        return userRepository.findById(id);
+    }
+
+    public Mono<User> save(User user) {
+        return userRepository.save(user);
+    }
+
+    public Mono<User> update(Long id, User user) {
```
Comment (FanilBiktimirov) [inline src/main/java/ru/jma/userservice/service/UserService.java]: Remove redundant parametres Long id

File: src/test/java/ru/jma/userservice/UserControllerTest.java [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+
+@WebFluxTest(UserController.class)
+public class UserControllerTest {
+
+    @Autowired
+    WebTestClient webTestClient;
+
+    @MockBean
+    UserService userService;
+
```
Comment (FanilBiktimirov) [inline src/test/java/ru/jma/userservice/UserControllerTest.java:33]: Add test data in head of class. With annotation @BeforeAll

File: src/test/java/ru/jma/userservice/UserControllerTest.java [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+    @Test
+    void testGetUser() {
+        User user = new User((long) 1, "user", "1234");
+        User user2 = new User((long) 2, "user2", "5678");
+        Flux<User> userFlux = Flux.just(user, user2);
+        Mockito.when(userService.findAll()).thenReturn(userFlux);
```
Comment (FanilBiktimirov) [inline src/test/java/ru/jma/userservice/UserControllerTest.java]: Add check array size

**Review by FanilBiktimirov — CHANGES_REQUESTED** [review]

File: src/test/java/ru/jma/userservice/UserControllerTest.java (hunk 12 above)
Comment (Johnny7c3) [reply src/test/java/ru/jma/userservice/UserControllerTest.java]: The WebTestClient interface does not contain a hasSize() method unlike the MockMvc Class

*WebTestClient: https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/test/web/reactive/server/WebTestClient.html
*MockMvc: https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/test/web/servlet/MockMvc.html

File: src/main/java/ru/jma/userservice/model/User.java (hunk 3 above)
Comment (Johnny7c3) [reply src/main/java/ru/jma/userservice/model/User.java:20]: The @AllArgsConstructor annotation is needed for the @With annotation to work, which is used in the UserService service

File: src/main/java/ru/jma/userservice/service/UserService.java (hunk 10 above)
Comment (Johnny7c3) [reply src/main/java/ru/jma/userservice/service/UserService.java]: Without this parameter, updating of a specific user will not work correctly

File: src/main/java/ru/jma/userservice/controller/UserController.java [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+
+@RestController
+@RequestMapping("/users")
+@RequiredArgsConstructor
+public class UserController {
+
+    private final UserService userService;
+
+    @GetMapping("/")
+    public Flux<User> getAllUsers() {
```
Comment (FanilBiktimirov) [inline src/main/java/ru/jma/userservice/controller/UserController.java:18]: take away "/"

File: src/main/java/ru/jma/userservice/controller/UserController.java [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+        return userService.getAllUsers();
+    }
+
+    @PostMapping("/")
```
Comment (FanilBiktimirov) [inline src/main/java/ru/jma/userservice/controller/UserController.java]: take away "/"

**Review by FanilBiktimirov — CHANGES_REQUESTED** [review]

**Review by FanilBiktimirov — APPROVED** [review]

**Review by ilya2236 — APPROVED** [review]

**Pull request merged**